### ISSUE TITLE
track() slower for every input when first calls use numpy frames vs tensor warmup — steady-state latency differs (~18ms vs ~11ms)

### ISSUE THREAD

**Issue by EgorLeMeilleur** [ultralytics/ultralytics#21819]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

## Summary
When running a YOLOv11s-pose model exported to TensorRT FP16 on an RTX 4070, I observe that **every** `model.track()` call runs ~18 ms/frame if I immediately begin processing video frames (numpy lists from cv2). However, if I first call `model.track()` with a  `torch.Tensor` (dummy warmup), every subsequent call runs ~11 ms/frame. This behavior occurs for both `yolov11s` and `yolov11s-pose` engines and not only for TensorRT but all base models, simply put for every model when tracking. The `predict()` path does not show this discrepancy.

### Environment

Ultralytics 8.3.186 🚀 Python-3.10.18 torch-2.8.0+cu128 CUDA:0 (NVIDIA GeForce RTX 4070, 11870MiB)
Setup complete ✅ (16 CPUs, 31.1 GB RAM, 784.5/914.8 GB disk)

OS                     Linux-6.14.0-28-generic-x86_64-with-glibc2.39
Environment            Linux
Python                 3.10.18
Install                pip
Path                   /home/gpokumin/Desktop/ml_work/venv10/lib/python3.10/site-packages/ultralytics
RAM                    31.12 GB
Disk                   784.5/914.8 GB
CPU                    Intel Core(TM) i5-14400
CPU count              16
GPU                    NVIDIA GeForce RTX 4070, 11870MiB
GPU count              1
CUDA                   12.8

numpy                  ✅ 2.2.6>=1.23.0
matplotlib             ✅ 3.10.5>=3.3.0
opencv-python          ✅ 4.12.0.88>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.2>=5.3.1
requests               ✅ 2.32.4>=2.23.0
scipy                  ✅ 1.15.3>=1.4.1
torch                  ✅ 2.8.0>=1.8.0
torch                  ✅ 2.8.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.23.0>=0.9.0
psutil                 ✅ 7.0.0
py-cpuinfo             ✅ 9.0.0
pandas                 ✅ 2.3.1>=1.1.4
ultralytics-thop       ✅ 2.0.15>=2.0.0

## First run without dummy input
```python
from ultralytics import YOLO
import time
import cv2

model = YOLO("yolo11s.pt")

video_path = "path/to/video.mp4"
cap = cv2.VideoCapture(video_path)

yolo_work_time = []
while cap.isOpened():
    success, frame = cap.read()

    if success:
        start = time.time()
        results = model.track(frame, persist=True)
        yolo_work_time.append(time.time() - start)
        annotated_frame = results[0].plot()

        cv2.imshow("YOLO11 Tracking", 
                   annotated_frame)

        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        break

cap.release()
cv2.destroyAllWindows()

print(sum(yolo_work_time))
print(sum(yolo_work_time)/len(yolo_work_time))
```
## Second run with dummy input
```python
from ultralytics import YOLO
import time
import torch
import cv2

model = YOLO("yolo11s.pt")

video_path = "path/to/video.mp4"
cap = cv2.VideoCapture(video_path)

batch = torch.rand([16, 3, 640, 640], dtype=torch.float32).to('cuda')
results_all = model.track(batch, persist=True, conf=0.75, verbose=False)

yolo_work_time = []
while cap.isOpened():
    success, frame = cap.read()

    if success:
        start = time.time()
        results = model.track(frame, persist=True)
        yolo_work_time.append(time.time() - start)
        annotated_frame = results[0].plot()

        cv2.imshow("YOLO11 Tracking", 
                   annotated_frame)

        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        break

cap.release()
cv2.destroyAllWindows()

print(sum(yolo_work_time))
print(sum(yolo_work_time)/len(yolo_work_time))
```
## My results
Tested on 5 minute 25 fps video in mp4 format.
Case 1 (no tensor warmup): ~168 seconds (0.0226s on average per frame)
Case 2 (tensor warmup): ~100 seconds (0.0135s on average per frame)

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @EgorLeMeilleur, thank you for your detailed report and interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will review and assist you shortly.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. Thank you for already sharing a clear script and timings. To help us narrow this down further, please also include:
- Exact model artifacts used (e.g., .pt vs TensorRT engine path), and confirm if both YOLO11 and YOLO11-pose show the same behavior for you
- Any tracker-specific settings beyond `persist=True` (e.g., custom tracker YAML) and whether changing them affects timings
- A short run on a few frames that reproduces the latency gap along with the printed per-frame averages
- Confirmation that the behavior persists after upgrading to the latest `ultralytics` and testing in one of the verified environments below
- Whether forcing `device='cuda'` in `model.track(..., device=0)` changes the results when starting from numpy frames

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by RizwanMunawar** [comment]
@EgorLeMeilleur, I think it's completely fine. Becuase whenever you run object tracking, on the first frame, it always takes time for warmup. If you warm up the tracker earlier, as you did in the second run, and pass the frame, the tracker will immediately start tracking it. Note: It also depends on the device, i.e, if you are running first time after a device restart, CUDA modules loading may take more time, and it will also affect the tracker speed.

Edit: I will investigate it further; maybe we can improve the tracker initialization a bit. Thanks

**Comment by Y-T-G** [comment]
Do you have the verbose logs from Ultralytics from both runs?

Can you post a snippet from both?

**Comment by EgorLeMeilleur** [comment]
> Do you have the verbose logs from Ultralytics from both runs?
> 
> Can you post a snippet from both?

The verbose logs are nearly identical between the two experiments, because it only prints detection/pose timings — not tracker work.

## Example snippets:

### No warmup:
```
0: 384x640 4 cows, 30.3ms
Speed: 2.6ms preprocess, 30.3ms inference, 35.3ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 4 cows, 3.8ms
Speed: 1.9ms preprocess, 3.8ms inference, 1.9ms postprocess per image at shape (1, 3, 384, 640)
```
### With warmup:
```
0: 640x640 (no detections), 3.2ms
1: 640x640 (no detections), 3.2ms
2: 640x640 (no detections), 3.2ms
3: 640x640 (no detections), 3.2ms
4: 640x640 (no detections), 3.2ms
5: 640x640 (no detections), 3.2ms
6: 640x640 (no detections), 3.2ms
7: 640x640 (no detections), 3.2ms
8: 640x640 (no detections), 3.2ms
9: 640x640 (no detections), 3.2ms
10: 640x640 (no detections), 3.2ms
11: 640x640 (no detections), 3.2ms
12: 640x640 (no detections), 3.2ms
13: 640x640 (no detections), 3.2ms
14: 640x640 (no detections), 3.2ms
15: 640x640 (no detections), 3.2ms
Speed: 0.0ms preprocess, 3.2ms inference, 1.3ms postprocess per image at shape (1, 3, 640, 640)

0: 384x640 9 cows, 33.6ms
Speed: 1.9ms preprocess, 33.6ms inference, 20.0ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 4 cows, 6.3ms
Speed: 2.3ms preprocess, 6.3ms inference, 0.7ms postprocess per image at shape (1, 3, 384, 640)
```
There are no further differences printed because the tracking work is not included in these logs:
```
0: 384x640 4 cows, 3.5ms Speed: 1.3ms preprocess, 3.5ms inference, 0.7ms postprocess per image at shape (1, 3, 384, 640)
```

**Comment by Y-T-G** [comment]
I believe it's because the warmup you do breaks `gmc` for BoTSORT, which is one of the techniques that consume latency. You can try testing with `tracker="bytetrack.yaml"` or by setting `gmcMethod: None` in `botsort.yaml`.
